### PR TITLE
Delete recorded consumers when autorecovering model is disposed

### LINKED ISSUE #1302: Consumer tags aren't cleaned on channel close causing memory leak.
RabbitMQ Client Library Version: C# 6.4
RabbitMQ Server Version: NA
Erlang Version: NA
Operating System: NA
RabbitMQ Plugins: NA

We are using pooling to keep 20 connections open to RabbitMQ server. Those 20 connections are used to open channels and to start consumers. We can have from 500 to 1000 consumers spread across those 20 connections. Most of them idle. We open a new channel and consumer to Rabbit MQ when one of our customers opens a new TCP connection to us. When they disconnect from us we in turn disconnect from Rabbit MQ server. Some of our customers regularly drop the connection and reconnect, every 10 to 15 minutes. This causes us to open and close thousands of channels a day.

When a new consumer is created, the call to RecordConsumer on the AutorecoveringConnection adds the consumer tag to an internal dictionary of consumer tags and RecordedConsumers. The item that is added can only be cleared out via the DeleteRecordedConsumer method which in turn is only called via a BasicCancel or BasicCancelNoWait. Simply closing the channel is not sufficient.

In our case, after running for a week we had over 53 000 RecordedConsumer objects. We took process dumps and analyzed the dumps for the memory leak.

### PR DESCRIPTION
**Pull request by lukebakken**
Fixes #1302

### CODE REVIEW

**Comment by lukebakken** [comment]
Hmm I appear to have broken something 😉

File: projects/RabbitMQ.Client/client/impl/AutorecoveringModel.cs [hunk 1]
```diff
@@ -1184,16 +1189,16 @@ public void BasicCancelNoWait(string consumerTag)
                 throw new ObjectDisposedException(GetType().FullName);
             }
 
-            string result = _delegate.BasicConsume(queue, autoAck, consumerTag, noLocal,
+            string resultConsumerTag = _delegate.BasicConsume(queue, autoAck, consumerTag, noLocal,
                 exclusive, arguments, consumer);
-            RecordedConsumer rc = new RecordedConsumer(this, queue).
-                WithConsumerTag(result).
+            RecordedConsumer rc = new RecordedConsumer(this, queue, resultConsumerTag).
```
Comment (lukebakken) [inline projects/RabbitMQ.Client/client/impl/AutorecoveringModel.cs:1194]: @Zerpet this is where I introduced the bug in a prior commit. I removed the code that sets the consumer tag, whoops. So, I changed the `RecordedConsumer` ctor to _require_ a tag.

File: projects/RabbitMQ.Client/client/impl/RecordedConsumer.cs [hunk 2]
```diff
@@ -29,16 +29,41 @@
...
         {
-            Model = model;
-            Queue = queue;
+            if (model == null)
+            {
+                throw new ArgumentNullException(nameof(model));
+            }
+            else
+            {
+                Model = model ?? throw new ArgumentNullException(nameof(model));
```
Comment (Zerpet) [inline projects/RabbitMQ.Client/client/impl/RecordedConsumer.cs:47]: Can `model` ever be `null` (or null-like) here after skiping the section `model == null` ?

**Comment by lukebakken** [comment]
Yep, I'm getting it right now. I can't even cherry-pick anymore since `main` is so different. Good incentive to get 7.0 out!

**Pull request merged**